packages/bvaughn-architecture-demo/src/contexts/SelectedFrameContext.tsx

```diff
@@ -0,0 +1,61 @@
+import { FrameId, PauseId } from "@replayio/protocol";
+import { createContext, PropsWithChildren, useContext, useEffect, useMemo, useState } from "react";
+import useLoadedRegions from "../hooks/useRegions";
+import { ReplayClientContext } from "shared/client/ReplayClientContext";
+import { TimelineContext } from "./TimelineContext";
+import { isPointInRegions } from "shared/utils/time";
+import { getPauseForExecutionPoint } from "../suspense/PauseCache";
+
+interface PauseAndFrameId {
+  pauseId: PauseId;
+  frameId: FrameId;
+}
+
+export interface SelectedFrameContextType {
+  selectedFrameId: PauseAndFrameId | null;
+  setSelectedFrameId(pauseAndFrameId: PauseAndFrameId | null): void;
+}
+
+export const SelectedFrameContext = createContext<SelectedFrameContextType>({
+  selectedFrameId: null,
+  setSelectedFrameId: () => {},
+});
+
+export function SelectedFrameContextRoot({ children }: PropsWithChildren<{}>) {
+  const [selectedFrameId, setSelectedFrameId] = useState<PauseAndFrameId | null>(null);
+  const context = useMemo(
+    () => ({ selectedFrameId, setSelectedFrameId }),
+    [selectedFrameId, setSelectedFrameId]
+  );
+  return <SelectedFrameContext.Provider value={context}>{children}</SelectedFrameContext.Provider>;
+}
+
+function SelectedFrameContextAdapter() {
+  const client = useContext(ReplayClientContext);
+  const loadedRegions = useLoadedRegions(client);
+  const { executionPoint } = useContext(TimelineContext);
+  const { setSelectedFrameId } = useContext(SelectedFrameContext);
+
+  const isLoaded = loadedRegions !== null && isPointInRegions(executionPoint, loadedRegions.loaded);
+
+  const pause = isLoaded ? getPauseForExecutionPoint(client, executionPoint) : undefined;
+
+  const pauseId = pause?.pauseId;
+  const frameId = pause?.stack?.[0] ?? null;
+
+  useEffect(() => {
+    const pauseAndFrameId = pauseId && frameId ? { pauseId, frameId } : null;
+    setSelectedFrameId(pauseAndFrameId);
+  }, [setSelectedFrameId, pauseId, frameId]);
+
+  return null;
+}
```

Comment:
Looks like `SelectedFrameContextAdapter` _suspends_ (when it calls `getPauseForExecutionPoint`) but then it uses an effect to update context values.

If we keep the _suspending_ behavior, then it seems like the effect is unnecessary (and we should add a `Suspense` wrapper to avoid it suspending the context and its entire subtree).

But if we want to avoid suspending we could add/use an async "helper" to the pause cache to load the data (so the effect loads and updates the context).

What are your thoughts?

Reply:
> If we keep the suspending behavior, then it seems like the effect is unnecessary

I added the effect because I thought render functions should not call a setter.

> and we should add a Suspense wrapper to avoid it suspending the context and its entire subtree

`SelectedFrameContextAdapter` doesn't take any `children`, so there is no subtree.

Why do you want to avoid suspending `SelectedFrameContextAdapter`?

Reply:
Suspending during render just to call an update function in an effect seems awkward to me.

Functions can call a state updater during render. It's not _ideal_ because it restarts some of the render work, but it can be done. (That's what you're doing in the other `SelectedFrameContextAdapter` component.)

Reply:
> Functions can call a state updater during render.

I didn't know that, thanks.
So I removed the effect and also added a guard so that we only call `setSelectedPauseAndFrameId()` if it really changed.
And I added a Suspense wrapper because I realized that's necessary (otherwise it would suspend the entire subtree of the next Suspense wrapper, not just its own (empty) subtree).